File: python_version/cli/commands/similar.py
import json
from typing import Any, Dict

from cli.utils.leetcode_api import get_problem_by_id, get_problem_details
from cli.utils.ui import (
    print_command_banner, print_success, print_error, print_info, print_warning,
    styled_text_input, print_section, console
)

def main(config: Dict[str, Any]):
    """Fetch and display similar problems for a LeetCode problem."""
    print_command_banner("Find Similar Problems")
    
    problem_num = styled_text_input("Enter problem number")
    
    print_info("Looking up problem...")
    problem_data = get_problem_by_id(problem_num)
    
    if not problem_data:
        print_error(f"Could not find problem with ID {problem_num}")
        return
        
    problem_name = problem_data["title"]
    slug = problem_data["slug"]
    
    print_success(f"Found: {problem_name}")
    print_info("Fetching similar problems...")
    
    details = get_problem_details(slug)
    
    if not details:
        print_error("Failed to fetch problem details from LeetCode.")
        return
        
    similar_json = details.get("similarQuestions")
    
    if not similar_json:
        print_warning("No similar problems available for this problem.")
        return
        
    try:
        similar_problems = json.loads(similar_json)
    except json.JSONDecodeError:
        print_error("Failed to parse similar problems data.")
        return
        
    if not similar_problems:
        print_warning("No similar problems found for this problem.")
        return
        
    print_section(f"Similar Problems to {problem_name} ({len(similar_problems)} total)")
    
    for i, prob in enumerate(similar_problems, 1):
        diff_color = {
            "Easy": "green",
            "Medium": "yellow",
            "Hard": "red"
        }.get(prob.get("difficulty"), "white")
        
        console.print(f"  [bold cyan]{i}.[/bold cyan] [white]{prob.get('title')}[/white] - [{diff_color}]{prob.get('difficulty')}[/{diff_color}] [dim](Slug: {prob.get('titleSlug')})[/dim]")
        
    console.print()
    print_success("Finished showing similar problems.")

File: go-cli/internal/commands/similar.go
package commands

import (
	"encoding/json"
	"strings"

	"leetcli/internal/api"
	"leetcli/internal/config"
	"leetcli/internal/template"
)

type SimilarProblem struct {
	Title       string `json:"title"`
	TitleSlug   string `json:"titleSlug"`
	Difficulty  string `json:"difficulty"`
}

func Similar(args []string, cfg *config.Config, ui UI) {
	ui.WriteOutput(MsgPlain, "--- Find Similar Problems ---\n")

	problemNum := ui.PromptText("Enter problem number")
	if problemNum == "" {
		return
	}

	ui.WriteOutput(MsgInfo, "Looking up problem...")
	problemData, err := api.GetProblemByID(problemNum)
	if err != nil {
		ui.WriteOutput(MsgError, "Could not find problem with ID %s", problemNum)
		return
	}

	ui.WriteOutput(MsgSuccess, "Found: %s", problemData.Title)
	ui.WriteOutput(MsgInfo, "Fetching similar problems...")

	details, err := api.GetProblemDetails(problemData.Slug)
	if err != nil {
		ui.WriteOutput(MsgError, "Failed to fetch problem details from LeetCode.")
		return
	}

	if details.SimilarQuestions == "" {
		ui.WriteOutput(MsgInfo, "No similar problems available for this problem.")
		return
	}

	var similar []SimilarProblem
	if err := json.Unmarshal([]byte(details.SimilarQuestions), &similar); err != nil {
		ui.WriteOutput(MsgError, "Failed to parse similar problems data.")
		return
	}

	if len(similar) == 0 {
		ui.WriteOutput(MsgInfo, "No similar problems found for this problem.")
		return
	}

	languages := template.NormalizeLanguages(nil)
	for k, v := range cfg.Languages {
		languages[k] = template.LanguageInfo{Label: v.Label, Ext: v.Ext}
	}
	var exts []string
	for _, info := range languages {
		exts = append(exts, info.Ext)
	}
	solvedMap := template.GetSolvedMap(cfg.BaseDir, exts)

	ui.WriteOutput(MsgPlain, "\nSimilar Problems to %s (%d total):", problemData.Title, len(similar))

	for i, sp := range similar {
		difficulty := sp.Difficulty
		badge := "[○ Unsolved]"
		if solvedMap[sp.TitleSlug] || solvedMap[api.Slugify(sp.Title)] {
			badge = "[✔ Solved]"
		}
		ui.WriteOutput(MsgPlain, "  %d. %-35s (%s) %s", i+1, sp.Title, difficulty, badge)
	}

	ui.WriteOutput(MsgSuccess, "Finished showing similar problems.")
}

var _ = strings.Join
var _ = template.CountSolutions
